### PR TITLE
Fixing a file consistency issue on SFTP upload

### PR DESCRIPTION
**Pull request by zveinn**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Our community raised an [issue with SFTP](https://github.com/minio/minio/issues/18415) where files would randomly get corrupted on upload. The issue is that SFTP demands underlying implementations to offer multiplex support for file segments, while our SFTP to S3 interface was assuming writes would be in sync. 

## Resolution
In order to ensure file consistency, a file segment queue was created which ensures that our S3 store receives all file segments in order. 

## NOTES
The current queue size is set to 1000 segments, if filled the writer returns an error "sftp segment queue is full".
During testing using local interfaces I would regularly run into a queue size of 1-10, so it is probably safe to assume that under certain circumstances the queue size can grow quite a bit bigger than 10. 

## How to test this PR?
The SFTP module relies on manual testing. I recommend uploading 1GB+ files.

1. Create a 1GB file
2. Get the md5 sum of the file
3. Upload the file and download it again
4. compare the md5sums of the downloaded file and the original

## Types of changes
- [ x ] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Review by harshavardhana — APPROVED** [review]

File: cmd/sftp-server-driver.go [hunk 1]
```diff
@@ -201,19 +201,83 @@ func (f *sftpDriver) Fileread(r *sftp.Request) (ra io.ReaderAt, err error) {
 	return obj, nil
 }
 
+func (w *writerAt) Close() error {
+	for i := range w.buffers {
```
Comment (klauspost) [inline cmd/sftp-server-driver.go]: Check if w.remaining > 0 - return error if so - and use w.w.CloseWithError.

File: cmd/sftp-server-driver.go [hunk 2]
```diff
@@ -201,19 +201,83 @@ func (f *sftpDriver) Fileread(r *sftp.Request) (ra io.ReaderAt, err error) {
...
+				return err
+			}
+			if n != len(w.buffers[i]) {
+				return fmt.Errorf("expected write size %d but wrote %d bytes", len(w.buffers[i]), n)
+			}
+			w.nextOffset += int64(len(w.buffers[i]))
+			w.offsets[i] = 0
+			w.buffers[i] = nil
+			w.remaining--
+			continue
```
Comment (klauspost) [inline cmd/sftp-server-driver.go]: You can have unflushed buffer at a lower index, that matches new offset. Either restart the loop or recurse.

If not, there is a risk you don't end flushing everything.

**Review by klauspost — CHANGES_REQUESTED** [review]

File: cmd/sftp-server-driver.go [hunk 3]
```diff
@@ -201,19 +201,83 @@ func (f *sftpDriver) Fileread(r *sftp.Request) (ra io.ReaderAt, err error) {
...
+	}
+
+	if w.remaining > 0 {
+		err = w.flushBuffers()
+		if err != nil {
+			return 0, err
+		}
+	}
+	return
 }
```
Comment (harshavardhana) [inline cmd/sftp-server-driver.go:256]: Now that this has gotten a bit complicated @zveinn can you add tests for boundary conditions?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/sftp-server-driver.go (hunk 2 above)
Comment (zveinn) [reply cmd/sftp-server-driver.go]: We actually ran into that problem during our first round of testing.

There are two options here, one is to keep iterating the buffer until we have no offsets left that fit the order OR we can trigger a flush anytime there is anything left in the buffer. Both will result in the full file be written. 

I opted to implement it this way because I felt like it was simpler to just flush anytime there was something in the buffer. Also, there is less time between accepting packets from the SFTP clients that way.. kind of like "1 packet in, 1x flush iteration".

File: cmd/sftp-server-driver.go (hunk 1 above)
Comment (zveinn) [reply cmd/sftp-server-driver.go]: will do

File: cmd/sftp-server-driver.go (hunk 3 above)
Comment (zveinn) [reply cmd/sftp-server-driver.go:256]: yep, I'll write up some tests for the writer after the weekend (or during if I find some time)

File: cmd/sftp-server-driver.go (hunk 2 above)
Comment (zveinn) [reply cmd/sftp-server-driver.go]: Now that I think about it.. my way of doing it is probably going to scale worse then trying to evict as many as possible on every flush. 

I'll have another look at this.

Comment (zveinn) [reply cmd/sftp-server-driver.go]: you're actually right and I can't believe I missed that.. I'll fix it asap

**Comment by zveinn** [comment]
I took another look at the problem and realized there was no need for a slice buffer. Please re-view the new map method. 

Will make some tests tomorrow  @klauspost @krisis @harshavardhana

**Review by krisis — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by klauspost** [comment]
@zveinn This is much better. Did want to ask for a refactor, since we could always add that later - but this is perfect!
